domain/watcher.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package domain
+
+import (
+	"sync"
+
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/changestream"
+	"github.com/juju/juju/core/watcher"
+	"github.com/juju/juju/core/watcher/eventsource"
+)
+
+type WatchableDBFactory = func() (changestream.WatchableDB, error)
+
+type WatcherFactory struct {
+	mu sync.Mutex
+
+	getDB       WatchableDBFactory
+	watchableDB changestream.WatchableDB
```

Comment:
Delete this, we don't want to cache this. It's already cached on the other side of `getDB`, so double caching can lead to bad outcomes.